cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMaker2AssemblyOperator.java

```diff
@@ -0,0 +1,374 @@
+        String mirrorMaker2Name = kafkaMirrorMaker2.getMetadata().getName();
+        if (kafkaMirrorMaker2.getSpec() == null) {
+            return maybeUpdateMirrorMaker2Status(reconciliation, kafkaMirrorMaker2,
+                    new InvalidResourceException("spec property is required"));
+        }
+        List<KafkaMirrorMaker2MirrorSpec> mirrors = Optional.ofNullable(kafkaMirrorMaker2.getSpec().getMirrors())
+                .orElse(Collections.emptyList());
+        String host = getServiceName(mirrorMaker2Name);
+        KafkaConnectApi apiClient = getKafkaConnectApi();
+        return apiClient.list(host, KafkaConnectCluster.REST_API_PORT).compose(deleteMirrorMaker2ConnectorNames -> {
+
+            for (Map.Entry<String, Function<KafkaMirrorMaker2MirrorSpec, KafkaConnectorSpec>> connectorEntry : MIRRORMAKER2_CONNECTORS.entrySet()) {
+                deleteMirrorMaker2ConnectorNames.removeAll(mirrors.stream()
+                        .filter(mirror -> connectorEntry.getValue().apply(mirror) != null) // filter out non-existent connectors
+                        .map(mirror -> mirror.getSourceCluster() + "->" + mirror.getTargetCluster() + connectorEntry.getKey())
+                        .collect(Collectors.toSet()));
+            }
+            log.debug("{}: {}} cluster: delete MirrorMaker 2.0 connectors: {}", reconciliation, kind(), deleteMirrorMaker2ConnectorNames);
+            Stream<Future<Void>> deletionFutures = deleteMirrorMaker2ConnectorNames.stream()
+                    .map(connectorName -> apiClient.delete(host, KafkaConnectCluster.REST_API_PORT, connectorName));
+            Stream<Future<Void>> createUpdateFutures = mirrors.stream()
+                    .map(mirror -> reconcileMirrorMaker2Connectors(reconciliation, host, apiClient, kafkaMirrorMaker2, mirror, mirrorMaker2Cluster));
+            return CompositeFuture.join(Stream.concat(deletionFutures, createUpdateFutures).collect(Collectors.toList())).map((Void) null);
+        });
+    }
+
+    private Future<Void> reconcileMirrorMaker2Connectors(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, KafkaMirrorMaker2 mirrorMaker2, KafkaMirrorMaker2MirrorSpec mirror, KafkaMirrorMaker2Cluster mirrorMaker2Cluster) {
+        String targetClusterAlias = mirror.getTargetCluster();
+        String sourceClusterAlias = mirror.getSourceCluster();
+        if (targetClusterAlias == null) {
+            return maybeUpdateMirrorMaker2Status(reconciliation, mirrorMaker2,
+                    new InvalidResourceException("targetCluster property is required"));
+        } else if (sourceClusterAlias == null) {
+            return maybeUpdateMirrorMaker2Status(reconciliation, mirrorMaker2,
+                    new InvalidResourceException("sourceCluster property is required"));
+        }
+        List<KafkaMirrorMaker2ClusterSpec> clusters = Optional.ofNullable(mirrorMaker2.getSpec().getClusters())
+            .orElse(Collections.emptyList());
```

Comment:
You're using this `ofNullable(List).orElse(emptyList)` pattern a lot. Perhaps we should add a helper method to `ModelUtils` for doing this.

Reply:
Done